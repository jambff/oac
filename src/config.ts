export type OpenApiClientEnv = 'staging' | 'production';

type BaseUrls = {
  staging: string;
  production: string;
};

const BASE_URLS: BaseUrls = {
  staging: 'https://example.staging-api.co.uk',
  production: 'https://example.api.co.uk',
};

export const getBaseUrl = (env?: OpenApiClientEnv, baseURL?: string) => {
  if (baseURL) {
    return baseURL;
  }

  if (!env) {
    throw new Error('Either a `baseURL` or an `env` must be given');
  }

  if (!(env in BASE_URLS)) {
    throw new Error(`Not a known \`env\`: ${env}`);
  }

  return BASE_URLS[env];
};
