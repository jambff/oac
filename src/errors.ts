type ValidationError = {
  property: string;
  constraint: string;
  message: string;
};

export class OpenApiClientError extends Error {
  statusCode: number;

  errors?: ValidationError[];

  isOpenApiClientError = true;

  constructor(statusCode: number, message: string, errors?: ValidationError[]) {
    super(
      `${message}${
        errors?.length
          ? ` ${errors
              ?.map(
                (err) => `${err.property} (${err.constraint}) ${err.message}`,
              )
              .join(' | ')}`
          : ''
      }`,
    );

    this.statusCode = statusCode;
    this.name = 'OpenApiClientError';
    this.errors = errors;
  }
}

/**
 * Determines whether a value is (probably) an error thrown by the API client.
 */
export const isOpenApiClientError = (val: any): val is OpenApiClientError =>
  !!val?.isOpenApiClientError;
