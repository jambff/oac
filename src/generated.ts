// The files referenced below are generated as part of the build process.
export {
  create,
  JamBffClient,
  JamBffRequest,
  OperationConfig,
  OperationOptions,
  // @ts-ignore
} from './client';
// @ts-ignore
export { version } from './version';
