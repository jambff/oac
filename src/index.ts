// @ts-ignore
import type { components } from './types';

// @ts-ignore
export * from './client';
export { isJamBffError, JamBffError } from './errors';
export { createJamBffClient, JamBffClientOptions } from './create';

export type JamBffModels = components['schemas'];
