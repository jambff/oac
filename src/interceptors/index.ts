export { createRefreshTokenInterceptor } from './refresh-access-token-interceptor';
export { createEconnresetInterceptor } from './econnreset-interceptor';
export { createResponseDebugInterceptor } from './response-debug-interceptor';
export { createUpgradeRequiredInterceptor } from './upgrade-required-interceptor';
