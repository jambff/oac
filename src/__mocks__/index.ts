// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { create, OpenApiRequest } from '../client';
import { OpenApiClientError } from '../errors';

const noop = {} as OpenApiRequest;

const operations = create(noop);

const mockClient = Object.keys(operations).reduce(
  (acc, operation) => ({
    ...acc,
    // @ts-ignore
    [operation]: jest.fn(() => {
      console.warn(
        `No mock return value set for operation ${operation}. ` +
          'Try adding a mock resolved value, for example: ' +
          `\`(openapiClient.${operation} as jest.Mock).mockResolvedValue({ foo: 'bar' })\``,
      );

      throw new OpenApiClientError(404, 'Not Found');
    }),
  }),
  {},
);

export { OpenApiClientError };

// Create a mock OpenAPI client.
// @ts-ignore
export const createOpenApiClient = jest.fn(() => mockClient);

// Use the real isOpenApiClientError function.
// @ts-ignore
export const isOpenApiClientError = jest.fn((err) =>
  // @ts-ignore
  jest.requireActual('../errors').isOpenApiClientError(err),
);
