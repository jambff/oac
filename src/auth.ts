import decode from 'jwt-decode';
import { JamBffError } from './errors';

type DecodedToken = {
  role?: string | string[];
  exp: number;
}

export type AccessToken = string | null | undefined;

export type TokenRetrieverFunction = () => AccessToken | Promise<AccessToken>;

/**
 * Get a property from the access token.
 */
const getTokenProperty = (accessToken: string, key: keyof DecodedToken) => {
  try {
    return (decode(accessToken) as DecodedToken)[key];
  } catch (err) {
    // Failed to decode token, ignore.
    return null;
  }
};

/**
 * Check if an access token has expired.
 */
const isTokenExpired = (accessToken: string) => {
  const exp = getTokenProperty(accessToken, 'exp');
  const timeNow = new Date().getTime() / 1000;

  return !exp || exp < timeNow;
};

/**
 * Throw a 401 Unauthorized error.
 */
const throwUnauthorizedError = (message: string) => {
  throw new JamBffError(401, message);
};

function assertDefined<T>(
  errorMessage: string,
  value?: T | null,
): asserts value is T {
  if (!value) {
    throwUnauthorizedError(errorMessage);
  }
}

const getRefreshedAccessToken = async (
  refreshAccessToken?: TokenRetrieverFunction,
): Promise<string> => {
  assertDefined(
    'Authorization is required but there is no valid access token and no `refreshAccessToken()` function was provided.',
    refreshAccessToken,
  );

  const refreshedToken = await refreshAccessToken();

  assertDefined(
    'Authorization is required but there is no valid access token and nothing was returned from `refreshAccessToken()`.',
    refreshedToken,
  );

  return refreshedToken;
};

/**
 * Get the bearer authorization header.
 */
export const getAuthorizationHeader = async (
  secure: boolean,
  getAccessToken?: TokenRetrieverFunction,
  refreshAccessToken?: TokenRetrieverFunction,
) => {
  if (secure) {
    assertDefined(
      'Authorization is required but no `getAccessToken()` function was provided.',
      getAccessToken,
    );
  }

  const noop = () => null;
  let accessToken = await (getAccessToken || noop)();

  if (!secure) {
    return null;
  }

  if (!accessToken) {
    accessToken = await getRefreshedAccessToken(refreshAccessToken);
  }

  // If the user is an admin and their token has expired but the route was not
  // secure anyway then we can still go ahead and make the request.
  if (!secure && isTokenExpired(accessToken)) {
    return null;
  }

  if (isTokenExpired(accessToken)) {
    accessToken = await getRefreshedAccessToken(refreshAccessToken);

    if (isTokenExpired(accessToken)) {
      throwUnauthorizedError(
        'Authorization is required but the access token has expired and `refreshAccessToken()` also returned an expired token.',
      );
    }
  }

  return `Bearer ${accessToken}`;
};
