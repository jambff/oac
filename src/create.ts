import axios, { AxiosInstance } from 'axios';
import qs from 'qs';
import { create, OpenApiClient, version } from './generated';
import { createRequestFunction } from './request';
import {
  createRefreshTokenInterceptor,
  createEconnresetInterceptor,
  createResponseDebugInterceptor,
  createUpgradeRequiredInterceptor,
  createRequestDebugInterceptor,
} from './interceptors';
import { TokenRetrieverFunction } from './auth';

export type OpenApiClientOptions = {
  baseURL: string;
  getAccessToken?: TokenRetrieverFunction;
  refreshAccessToken?: TokenRetrieverFunction;
  onError?: (error: any) => void;
  onUpgradeRequired?: () => void;
  debug?: boolean;
};

/**
 * Create the base axios instance.
 */
const createAxiosInstance = ({
  baseURL,
  refreshAccessToken,
  onError,
  onUpgradeRequired,
  debug,
}: OpenApiClientOptions): AxiosInstance => {
  if (!baseURL) {
    throw new Error('A `baseURL` must be given');
  }

  const axiosInstance = axios.create({
    baseURL,
    headers: {
      'Content-Type': 'application/json',
      Accept: `application/vnd.jambff+json; version=${version}`,
    },
    paramsSerializer: (params) =>
      qs.stringify(params, {
        encodeValuesOnly: true,
        arrayFormat: 'brackets',
      }),
  });

  const refreshTokenInterceptor =
    createRefreshTokenInterceptor(refreshAccessToken);

  const econnresetInterceptor = createEconnresetInterceptor();
  const responseDebugInterceptor = createResponseDebugInterceptor(
    onError,
    debug,
  );

  const requestDebugInterceptor = createRequestDebugInterceptor();
  const upgradeRequiredInterceptor =
    createUpgradeRequiredInterceptor(onUpgradeRequired);

  axiosInstance.interceptors.response.use(
    refreshTokenInterceptor.success,
    refreshTokenInterceptor.error,
  );

  axiosInstance.interceptors.response.use(
    econnresetInterceptor.success,
    econnresetInterceptor.error,
  );

  axiosInstance.interceptors.response.use(
    responseDebugInterceptor.success,
    responseDebugInterceptor.error,
  );

  if (onUpgradeRequired) {
    axiosInstance.interceptors.response.use(
      upgradeRequiredInterceptor.success,
      upgradeRequiredInterceptor.error,
    );
  }

  if (debug) {
    axiosInstance.interceptors.request.use(
      requestDebugInterceptor.success,
      requestDebugInterceptor.error,
    );
  }

  return axiosInstance;
};

/**
 * Create the API client.
 */
export const createOpenApiClient = (
  options: OpenApiClientOptions,
): OpenApiClient => {
  const axiosInstance = createAxiosInstance(options);
  const request = createRequestFunction(
    axiosInstance,
    options.getAccessToken,
    options.refreshAccessToken,
  );

  return create(request);
};
