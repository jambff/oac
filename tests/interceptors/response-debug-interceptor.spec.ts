import { AxiosError, AxiosResponse } from 'axios';
import { OpenApiClientError } from '../../src/errors';
import { createResponseDebugInterceptor } from '../../src/interceptors';

const originalConsoleWarn = console.warn;
const originalConsoleError = console.error;

describe('createResponseDebugInterceptor', () => {
  beforeEach(() => {
    console.warn = originalConsoleWarn;
    console.error = originalConsoleError;
  });

  describe('error', () => {
    it.each([400, 404])('does not log a %s error', async (statusCode) => {
      console.error = jest.fn();

      const error = {
        response: {
          status: statusCode,
          data: {
            statusCode,
            message: 'Client did something wrong',
            name: 'ClientError',
            errors: [],
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'get',
        },
      } as unknown as AxiosError;

      const onError = jest.fn();
      const interceptor = createResponseDebugInterceptor(onError);

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        `${statusCode} Client did something wrong <GET http://api.com/endpoint>`,
      );

      expect((err as OpenApiClientError).statusCode).toBe(statusCode);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual([]);

      expect(onError).not.toHaveBeenCalled();
      expect(console.error).not.toHaveBeenCalled();
    });

    it('logs a 500 error', async () => {
      console.error = jest.fn();

      const error = {
        response: {
          status: 500,
          data: {
            statusCode: 500,
            message: 'Internal Server Error',
            name: 'BadRequest',
            errors: [],
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'get',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor();

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Internal Server Error <GET http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual([]);

      expect(console.error).toHaveBeenCalledTimes(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('logs a 400 error if in debug mode', async () => {
      console.error = jest.fn();

      const errors = [
        {
          constraint: 'isInt',
          message: 'id must be an integer number',
          property: 'id',
        },
        {
          constraint: 'isString',
          message: 'name must be a string',
          property: 'name',
        },
      ];

      const error = {
        response: {
          status: 400,
          data: {
            statusCode: 400,
            message: 'Bad Request',
            name: 'BadRequest',
            errors,
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'put',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor(undefined, true);

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '400 Bad Request <PUT http://api.com/endpoint> id (isInt) id must be an integer number | name (isString) name must be a string',
      );

      expect((err as OpenApiClientError).statusCode).toBe(400);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual(errors);

      expect(console.error).toHaveBeenCalledTimes(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('logs a 500 error using a custom error handler', async () => {
      console.error = jest.fn();

      const error = {
        response: {
          status: 500,
          data: {
            statusCode: 500,
            message: 'Internal Server Error',
            name: 'BadRequest',
            errors: [],
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'post',
        },
      } as unknown as AxiosError;

      const onError = jest.fn();
      const interceptor = createResponseDebugInterceptor(onError);

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Internal Server Error <POST http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual([]);

      expect(onError).toHaveBeenCalledTimes(1);
      expect(onError).toHaveBeenCalledWith(err);

      expect(console.error).not.toHaveBeenCalled();
    });

    it.each([undefined, {}, ''])(
      'logs an error when the API response data is "%s"',
      async (data) => {
        console.error = jest.fn();

        const error = {
          message: 'Bad thing',
          response: {
            status: 500,
            data,
          },
          config: {
            url: '/endpoint',
            baseURL: 'http://api.com',
            method: 'post',
          },
        } as unknown as AxiosError;

        const interceptor = createResponseDebugInterceptor();

        let err;

        try {
          interceptor.error(error);
        } catch (e) {
          err = e;
        }

        expect((err as OpenApiClientError).message).toBe(
          '500 Bad thing <POST http://api.com/endpoint>',
        );

        expect((err as OpenApiClientError).statusCode).toBe(500);
        expect((err as OpenApiClientError).name).toBe('OpenApiClientError');

        expect(console.error).toHaveBeenCalledTimes(1);
        expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
      },
    );

    it('logs an error when the response body does not exist', async () => {
      console.error = jest.fn();

      const error = {
        statusCode: 500,
        message: 'Bad thing',
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'get',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor();

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Bad thing <GET http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toBeUndefined();

      expect(console.error).toHaveBeenCalledTimes(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('success', () => {
    it('returns the response on success', () => {
      const mockResponse = { foo: 'bar' };
      const interceptor = createResponseDebugInterceptor();

      const result = interceptor.success(
        mockResponse as unknown as AxiosResponse,
      );

      expect(result).toEqual(mockResponse);
    });
  });
});
