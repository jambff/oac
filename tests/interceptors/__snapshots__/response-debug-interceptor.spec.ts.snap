// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResponseDebugInterceptor error logs a 400 error if in debug mode 1`] = `[OpenApiClientError: 400 Bad Request <PUT http://api.com/endpoint> id (isInt) id must be an integer number | name (isString) name must be a string]`;

exports[`createResponseDebugInterceptor error logs a 500 error 1`] = `[OpenApiClientError: 500 Internal Server Error <GET http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "{}" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "undefined" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the response body does not exist 1`] = `[OpenApiClientError: 500 Bad thing <GET http://api.com/endpoint>]`;
